ksqldb-rest-client/src/main/java/io/confluent/ksql/rest/client/KsqlTargetUtil.java

```diff
@@ -69,6 +71,63 @@ public static List<StreamedRow> toRows(final Buffer buff) {
     return rows;
   }
 
+  /**
+   * @param buff the buffer response that we get from the /query-stream endpoint
+   *             for KsqlMediaType.KSQL_V1_PROTOBUF
+   * @return a List of StreamedRow that includes the header row as its fist element,
+   *     and every subsequent element is a result row in the protobuf format
+   */
+  public static List<StreamedRow> toRowsFromProto(final Buffer buff) {
+    final List<StreamedRow> rows = new ArrayList<>();
+    boolean headerRead = false;
+
+    // split the buffer across the commas delimiting the individual rows
+    // Since, there would be commas in the header we are replacing
+    // },{"row": with },\n\t\n{"row": and then splitting on \n\t\n to avoid
+    // splitting on commas that might be more that 1 level deep
+    final List<Buffer> buffRows = Arrays
+            .stream(buff
+                    .toString()
+                    .replace("},{\"row\":", "},\n\t\n{\"row\":")
+                    .split("\n\t\n"))
+            .collect(Collectors.toList())
+            .stream()
+            .map(row -> toJsonMsg(Buffer.buffer(row), true))
+            .collect(Collectors.toList());
+
+    for (Buffer tidied: buffRows) {
+      // deserialize the first row (header) into a StreamedRow and add it to
+      // the list to return.
+      if (!headerRead) {
+        final StreamedRow row = deserialize(tidied, StreamedRow.class);
+        rows.add(row);
+        headerRead = true;
+      } else {
+        try {
+          // try to deserialize  every subsequent buffer row into a StreamedRow
+          // and add it to the list to return. These can be continuation tokens,
+          // consistency tokens, error messages, final messages or limit messages.
+          // We expect it to fail for the rowProtobuf objects which is handled below.
+          rows.add(deserialize(tidied, StreamedRow.class));
+        } catch (Exception e) {
+          // if it is none of the above, then deserialize it to a
+          // StreamedRow.DataRowProtobuf object.
+          // We can't directly deserialize into a StreamedRow like we did above
+          // because DataRowProtobuf's serialization is {"row":"CAESBlVTRVJfMA=="}
+          // whereas the serialization of a StreamedRow containing a DataRowProtobuf
+          // looks like {"rowProtobuf":{"row":"CAESBlVTRVJfMA=="}}. The Rest API
+          // produces {"row":"CAESBlVTRVJfMA=="}, so we have to do a bit of extra conversion
+          // here.
+          final DataRowProtobuf protoBytes =
```

Comment:
You mention that the Rest API produced the StreamdRow directly - what produces `DataRowProtobuf` and can we just standardize on one format? (i.e. why can't they all produce just `{"row":"CAESBlVTRVJfMA=="}`?

Reply:
>You mention that the Rest API produced the StreamdRow directly
not 100%. The REST API for proto produces `StreamedRow` for everything except the protobuf data rows. For the protobuf data rows, the `JsonQueryStreamResponseWriter` would produce the `DataRowProtobuf` which is an optional member of `StreamedRow` and gets serialized to `{"row":"CAESBlVTRVJfMA=="}`. 

Why is there this weird conditionality for protobuf data rows? If we just serialize a `StreamedRow` that contains a `DataRowProtobuf`, then it would look something like `{"rowProtobuf":{"row":"CAESBlVTRVJfMA=="}` which is not the public api that we want to abide by (unless we change our minds and loosen the restrictions on the api).

We can achieve the behavior where serializing a `StreamedRow` that contains protobuf row data yields something closer to `{"row":"CAESBlVTRVJfMA=="}`, but we can never get it to be the same because `row` is already taken up by the regular JSON rows. I hope this makes sense